patch-tracing-channel.js

```diff
@@ -0,0 +1,163 @@
+const ReflectApply = Reflect.apply;
+const PromiseReject = Promise.reject;
+const PromiseResolve = Promise.resolve;
+const PromisePrototypeThen = Promise.prototype.then;
+const ArrayPrototypeSplice = Array.prototype.splice;
+
+class TracingChannel {
+  constructor(nameOrChannels) {
+    if (typeof nameOrChannels === 'string') {
+      this.start = channel(`tracing:${nameOrChannels}:start`);
+      this.end = channel(`tracing:${nameOrChannels}:end`);
+      this.asyncStart = channel(`tracing:${nameOrChannels}:asyncStart`);
+      this.asyncEnd = channel(`tracing:${nameOrChannels}:asyncEnd`);
+      this.error = channel(`tracing:${nameOrChannels}:error`);
+    } else if (typeof nameOrChannels === 'object') {
+      const { start, end, asyncStart, asyncEnd, error } = nameOrChannels;
+
+      // assertChannel(start, 'nameOrChannels.start');
+      // assertChannel(end, 'nameOrChannels.end');
+      // assertChannel(asyncStart, 'nameOrChannels.asyncStart');
+      // assertChannel(asyncEnd, 'nameOrChannels.asyncEnd');
+      // assertChannel(error, 'nameOrChannels.error');
+
+      this.start = start;
+      this.end = end;
+      this.asyncStart = asyncStart;
+      this.asyncEnd = asyncEnd;
+      this.error = error;
+    } else {
+      throw new ERR_INVALID_ARG_TYPE('nameOrChannels',
+                                     ['string', 'object', 'Channel'],
+                                     nameOrChannels);
+    }
+  }
+
+  subscribe(handlers) {
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      if (this[name]) this[name].subscribe(handlers[name]);
+    }
+  }
+
+  unsubscribe(handlers) {
+    let done = true;
+
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      if (!(this[name] && this[name].unsubscribe(handlers[name]))) {
+        done = false;
+      }
+    }
+
+    return done;
+  }
+
+  traceSync(fn, context = {}, thisArg, ...args) {
+    const { start, end, error } = this;
+
+    return start.runStores(context, () => {
```

Comment:
Will also need to add `runStores` and `bindStore` to the `Channel` classes for this to work.